resolwe/flow/executors/run.py

```diff
@@ -0,0 +1,322 @@
+        obj, ndx = decoder.raw_decode(text)
+
+        if not isinstance(obj, dict):
+            raise ValueError()
+
+        text = text[ndx:].lstrip('\r\n')
+        yield obj
+
+
+class BaseFlowExecutor(object):
+    """Represents a workflow executor."""
+
+    exported_files_mapper = defaultdict(dict)
+
+    def __init__(self, *args, **kwargs):
+        """Initialize attributes."""
+        self.data_id = None
+        self.process = None
+        # Flag to mark current object as failed.
+        self.requirements = {}
+        self.resources = {}
+        self.redis = redis.StrictRedis(**settings.FLOW_EXECUTOR.get('REDIS_CONNECTION', {}))
+        self.manager_cookie = 1
+        self.queue_response_channel = '{}.{}'.format(settings['REDIS_CHANNEL_PAIR'][1], settings['Data_get']['id'])
+
+    def get_tools(self):
+        """Get tools paths."""
+        tools_paths = settings.FLOW_EXECUTOR_TOOLS_PATHS
+        return tools_paths
+
+    def start(self):
+        """Start process execution."""
+        pass
+
+    def run_script(self, script):
+        """Run process script."""
+        raise NotImplementedError('`run_script` function must be defined')
+
+    def end(self):
+        """End process execution."""
+        pass
+
+    def get_stdout(self):
+        """Get process' standard output."""
+        return self.stdout  # pylint: disable=no-member
+
+    def _send_manager_command(self, cmd, expect_reply=True, **kwargs):
+        """Send a properly formatted command to the manager."""
+        packet = {
+            'id': settings['Data_get']['id'],
+            'action': cmd,
+            'cookie': self.manager_cookie,
+        }
+        packet.update(kwargs)
+
+        # TODO what happens here if the push fails? we don't have any realistic recourse,
+        # so just let it explode and stop processing
+        queue_channel = settings['REDIS_CHANNEL_PAIR'][0]
+        try:
+            self.redis.rpush(queue_channel, json.dumps(packet))
+        except:  # pylint: disable=bare-except
+            logger.error("Error sending command to manager:\n\n%s", traceback.format_exc())
+            raise
+
+        if not expect_reply:
+            return
+
+        while True:
+            for _ in range(3):
+                response = self.redis.blpop(self.queue_response_channel, timeout=1)
+                if response:
+                    break
+            else:
+                # NOTE: If there's still no response after three seconds, the system is broken
+                # enough that it makes sense to give up; we're isolated here, so if the manager
+                # doesn't respond, we can't really do much more than just crash
+                raise RuntimeError('No response from the manager after 3 retries.')
+
+            _, item = response
+            packet = json.loads(item.decode('utf-8'))
+            if packet['cookie'] != self.manager_cookie:
+                continue
+            assert packet['result'] == 'OK'
```

Comment:
Please do not hardcode protocol values like this. All protocol-related constants should go into its own `protocol` module or something like this.

Reply:
@dblenkus, @kostko, do you recall where we said that the `protocols` module should live?